### PR TITLE
Fix type checking issue in map constructing Query expressions for string subtypes

### PR DESCRIPTION
**Pull request by SandaruJayawardana**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/37401

## Check List 
- [X] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [X] Added necessary tests
  - [X] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -5902,7 +5902,8 @@ private BType getQueryMapConstraintType(BType type, Location pos) {
             }
         } else if (type.tag == TypeTags.TUPLE) {
             List<BType> tupleTypeList = ((BTupleType) type).tupleTypes;
-            if (tupleTypeList.size() == 2 && TypeTags.isStringTypeTag(tupleTypeList.get(0).tag)) {
+            if (tupleTypeList.size() == 2 &&
+                    TypeTags.isStringTypeTag(Types.getReferredType(tupleTypeList.get(0)).tag)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5906]: Still won't work for subtypes of string, right?

```ballerina
type FooBar "foo"|"bar";

function test() {
    [FooBar, int|float][] list = [["foo", 1.4], ["bar", 2], ["foo", 3]];
    map<int|float> & readonly|error mp1 = map from var item in list select item; // incompatible type in select clause: expected [string,any|error], found '[FooBar,(int|float)]'
}
```

Comment (SandaruJayawardana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5906]: will check.

Comment (SandaruJayawardana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/37402/commits/fa3d9b61f3e78f304e6ae2495183e2d58e77af98.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -5896,20 +5896,41 @@ private BType getTypeOfTypeParameter(BType selectType, Location pos) {
...
             BArrayType arrayType = (BArrayType) type;
-            if (arrayType.state != BArrayState.OPEN && arrayType.size == 2 &&
-                    TypeTags.isStringTypeTag(arrayType.eType.tag)) {
+            if (arrayType.state != BArrayState.OPEN && arrayType.size == 2 && isString(arrayType.eType)) {
                 return arrayType.eType;
             }
         } else if (type.tag == TypeTags.TUPLE) {
             List<BType> tupleTypeList = ((BTupleType) type).tupleTypes;
-            if (tupleTypeList.size() == 2 && TypeTags.isStringTypeTag(tupleTypeList.get(0).tag)) {
+            if (tupleTypeList.size() == 2 && isString(tupleTypeList.get(0))) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5904]: Why can't we just do `org.wso2.ballerinalang.compiler.semantics.analyzer.Types#isAssignable(org.wso2.ballerinalang.compiler.semantics.model.types.BType, org.wso2.ballerinalang.compiler.semantics.model.types.BType)` with `org.wso2.ballerinalang.compiler.semantics.model.SymbolTable#stringType` as the target?

Comment (SandaruJayawardana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5904]: +1. Will update. Thanks.

Comment (SandaruJayawardana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/37402/commits/385f06c9c723b66f489d8f370eb38cb49333e481.

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-query-construct-type-negative.bal [hunk 3]
```diff
@@ -284,4 +284,41 @@ function testConstructingInvalidReadonlyMap() {
...
+
+type FooBar1 ("foo"|"bar"|int)[2];
+type FooBar2 ("foo"|"bar"|1)[2];
+type FooBar3 "foo"|"bar"|1;
+type FooBar4 "foo"|"bar"|float;
+type FooBar5 "foo"|"bar"|1;
+
+function testConstructingInvalidMaps() {
+    FooBar1[] list1 = [];
+    map<FooBar1>|error mp1 = map from var item in list1 select item;
```
[the remark this answers is not in the document]
Comment (SandaruJayawardana) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-query-construct-type-negative.bal:302]: That fix is not merged yet. We can update these tests in the patch release.

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-query-construct-type-negative.bal [hunk 4]
```diff
@@ -284,4 +284,41 @@ function testConstructingInvalidReadonlyMap() {
...
+    map<int[]> & readonly|error mp7 = map from var item in list select item;
+}
+
+type FooBar1 ("foo"|"bar"|int)[2];
+type FooBar2 ("foo"|"bar"|1)[2];
+type FooBar3 "foo"|"bar"|1;
+type FooBar4 "foo"|"bar"|float;
+type FooBar5 "foo"|"bar"|1;
+
+function testConstructingInvalidMaps() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-query-construct-type-negative.bal:300]: Better to include more details in the name similar to other functions.

**Review by MaryamZi — APPROVED** [review]

**Review by gimantha — DISMISSED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-query-construct-type-negative.bal (hunk 4 above)
Comment (SandaruJayawardana) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-query-construct-type-negative.bal]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/37402/commits/f09fe71ace924324a7680aa0c10053aed965ebb2.

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**